subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultPluginManager.java

```diff
@@ -102,19 +107,28 @@ public PluginContainer getPluginContainer() {
 
     @Override
     public void apply(PluginImplementation<?> plugin) {
-        doApply(plugin);
+        doApplyThroughBuildOperation(plugin);
     }
 
     public void apply(String pluginId) {
         PluginImplementation<?> plugin = pluginRegistry.lookup(DefaultPluginId.unvalidated(pluginId));
         if (plugin == null) {
             throw new UnknownPluginException("Plugin with id '" + pluginId + "' not found.");
         }
-        doApply(plugin);
+        doApplyThroughBuildOperation(plugin);
     }
 
     public void apply(Class<?> type) {
-        doApply(pluginRegistry.inspect(type));
+        doApplyThroughBuildOperation(pluginRegistry.inspect(type));
+    }
+
+    private void doApplyThroughBuildOperation(final PluginImplementation<?> plugin) {
+        new ApplyPluginBuildOperation(plugin, buildOperationExecutor).run(new Action<BuildOperationContext>() {
+            @Override
+            public void execute(BuildOperationContext buildOperationContext) {
+                doApply(plugin);
```

Comment:
We can't wrap all of `doApply`, because that checks whether the plugin has already been applied. Instead, we should just wrap the code path inside `doApply`, which runs when the plugin has not been added before.